OpenRA.Mods.Common/Traits/Conditions/GrantConditionAfterDelay.cs

```diff
@@ -0,0 +1,103 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2018 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Drawing;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Gives a condition to the actor after a delay.")]
+	public class GrantConditionAfterDelayInfo : PausableConditionalTraitInfo
+	{
+		[FieldLoader.Require]
+		[GrantedConditionReference]
+		[Desc("The condition to grant")]
+		public readonly string Condition = null;
+
+		[Desc("Number of ticks to wait before applying the condition.")]
+		public readonly int Delay = 25;
+
+		public readonly bool ShowSelectionBar = true;
+		public readonly bool ShowFullBarAfterGranted = true;
+		public readonly Color SelectionBarColor = Color.Magenta;
+
+		public override object Create(ActorInitializer init) { return new GrantConditionAfterDelay(this); }
+	}
+
+	public class GrantConditionAfterDelay : PausableConditionalTrait<GrantConditionAfterDelayInfo>, ITick, ISync, INotifyCreated, ISelectionBar
+	{
+		readonly GrantConditionAfterDelayInfo info;
+		ConditionManager manager;
+		int token = ConditionManager.InvalidConditionToken;
+		[Sync] public int Ticks { get; private set; }
+
+		public GrantConditionAfterDelay(GrantConditionAfterDelayInfo info)
+			: base(info)
+		{
+			this.info = info;
+			Ticks = info.Delay;
+		}
+
+		protected override void Created(Actor self)
+		{
+			manager = self.TraitOrDefault<ConditionManager>();
+
+			base.Created(self);
+		}
+
+		void GrantCondition(Actor self, string cond)
+		{
+			if (manager == null)
+				return;
+
+			if (string.IsNullOrEmpty(cond))
+				return;
+
+			token = manager.GrantCondition(self, cond);
+		}
+
+		void ITick.Tick(Actor self)
+		{
+			if (IsTraitDisabled)
+				Ticks = info.Delay;
+
+			if (IsTraitPaused || IsTraitDisabled)
+				return;
+
+			if (--Ticks < 0)
+				if (token == ConditionManager.InvalidConditionToken)
+					GrantCondition(self, info.Condition);
+		}
+
+		float ISelectionBar.GetValue()
+		{
+			if (IsTraitDisabled || !Info.ShowSelectionBar || ((1f - (float)Ticks / Info.Delay) > 1f && !info.ShowFullBarAfterGranted))
```

Comment:
This line is a bit of a monster. Would you mind splitting this into one check for `IsTraitDisabled || !Info.ShowSelectionBar ` and then a second for the rest?

The `1f - (float)Ticks / Info.Delay) > 1f` looks dodgy, but my comment above applies here too.

Reply:
That ShowFullBarAfterGranted was a hack, timer shouldn't go down once the condition is granted as you said below.